### DISCUSSION TITLE
How can I use thrust in Python

### DISCUSSION THREAD

**Discussion by BolunDai0216** [NVIDIA/cccl#1683] [Thrust]
Hi,

I was using thrust to create a library in C++, however, I also want to pybind the the C++ library and use it in Python. I was wondering if anyone has an example of how to do this? It seems like the process is a bit different compared to pybinding CPU-only libraries.

Also, is pybinding the library a good way to use it in Python or if there are better ways?

Thanks in advance!

**Comment by jrhemstad** [comment]
Hey @BolunDai0216!

Thanks for your interest in CCCL and Thrust! 

We work with a lot of Python folks, so we understand where you are coming from.

I think the main problem you'll run into with just trying to use pybind with Thrust is that Thrust is a C++ template library. So you don't know the types ahead of time you'll be working on, and you'd need some kind of runtime compilation approach.

Have you looked into [`CuPy`](https://docs.cupy.dev/en/stable/install.html)? They actually use Thrust and CUB internally for some of the functionality they provide. It works based on pre-instantiating Thrust templates for fixed types. 

Could you share more about what kinds of things you're trying to do?

**Reply by BolunDai0216** [reply]
@jrhemstad Thanks for the quick response!

I mostly want to use `thrust` to create custom kernels that I can use to process point clouds, like computing a custom distance metric between each point in the point cloud and a geometric shape.

I haven't tried CuPy before but I will take a look at it. Thanks for sharing this.

I can share some of the code after I clean it up to showcase what I am doing and the errors I am encountering.

**Reply by ikaryss** [reply]
@BolunDai0216 Hello!  
I am currently working on a similar problem, I want to use a custom thrust kernel in Python.  
I would like to know what solution you came to in your case?

**Reply by jrhemstad** [reply]
I've got some good news! Since this original post, we've started working on exposing both Thrust and CUB functionality to Python. 

It is still *very* experimental and early, but you can check out what we're working on here: https://nvidia.github.io/cccl/python.html

**Reply by BolunDai0216** [reply]
> @BolunDai0216 Hello!  
> 
> I am currently working on a similar problem, I want to use a custom thrust kernel in Python.  
> 
> I would like to know what solution you came to in your case?

@ikaryss unfortunately I did not pursue that route. I used warp to perform the functionalities instead.

**Reply by BolunDai0216** [reply]
> I've got some good news! Since this original post, we've started working on exposing both Thrust and CUB functionality to Python. 
> 
> 
> 
> It is still *very* experimental and early, but you can check out what we're working on here: https://nvidia.github.io/cccl/python.html 

@jrhemstad this is awesome! Really excited about this!
